src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -3137,6 +3137,55 @@ int32_t NaClSysImcSocketPair(struct NaClAppThread *natp,
   return retval;
 }
 
+int32_t NaClSysSocketPair(struct NaClAppThread *natp,
+                          int                  domain,
+                          int                  type,
+                          int                  protocol,
+                          int                  *fds) {
+
+  struct NaClApp          *nap = natp->nap;
+  void                    *hd_struct = NULL;
+  struct NaClHostDesc     *hd;
+  int                     lind_fd;
+  int                     nacl_fd;
+  int                     user_fd;
+  int32_t                 retval;
+
+  NaClLog(2, "Cage %d Entered NaClSysSocketPair(0x%08"NACL_PRIxPTR", "
+           "%d, %d, %d, 0x%08"NACL_PRIx32")\n",
+           nap->cage_id, (uintptr_t)natp, domain, type, protocol, (uintptr_t)fds);
+
+  //Preprocessing start
+  hd_struct = malloc(sizeof(struct NaClHostDesc)*2);
+  if (!hd_struct) {
+    retval = -NACL_ABI_ENOMEM;
+    return retval;
+  }
+  //Preprocessing end
+
+  retval = lind_socketpair (domain, type, protocol, fds, nap->cage_id);
+
+  //Postprocessing start
+  for(int i=0; i<2; ++i) {
+    hd = &((struct NaClHostDesc*)hd_struct)[i];
+    lind_fd = ((int*)fds)[i];
+
+    //NaClHostDescCtor(hd, lind_fd, NACL_ABI_O_RDWR):
+    hd->d = lind_fd;
+    hd->flags = NACL_ABI_O_RDWR;
+
+    hd->cageid = nap->cage_id;
+    nacl_fd = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescIoDescMake(hd)));
+    user_fd = NextFd(nap->cage_id);
+    fd_cage_table[nap->cage_id][user_fd] = nacl_fd;
+  }
+  //Postprocessing end
+
+  NaClLog(2, "NaClSysSocketPair: returning %d\n", retval);
+
```

Comment:
From what I understand you'll need to replace the 2 fds in the array with the system fds you create for them (what you assign to user_fd), so that the user program has the right fds to use